cpp/include/raft/distance/detail/distance.cuh

```diff
@@ -210,15 +210,16 @@ void distance_impl(raft::resources const& handle,
                 "OutT can be uint8_t, float, double,"
                 "if sizeof(OutT) > 1 then sizeof(AccT) == sizeof(OutT).");
 
-  ASSERT(!(((x != y) && (worksize < (m + n) * sizeof(AccT))) || (worksize < m * sizeof(AccT))),
+  ASSERT(!((((x != y) || ((x == y) && (m != n))) && (worksize < (m + n) * sizeof(AccT))) ||
+           (worksize < m * sizeof(AccT))),
          "workspace size error");
   ASSERT(workspace != nullptr, "workspace is null");
 
   cudaStream_t stream = raft::resource::get_cuda_stream(handle);
 
   DataT* x_norm = workspace;
   DataT* y_norm = workspace;
-  if (x != y) {
+  if ((x != y) || ((x == y) && (m != n))) {
```

Comment:
This solves the problem. But what do you think about an alternative solution?

```c++
if (x != y) {
// ... keep original code 
} else {
   // use max(m,n)
    raft::linalg::rowNorm(
      x_norm, x, k, std::max(m,n), raft::linalg::L2Norm, is_row_major, stream, raft::sqrt_op{});
  }
```
During workspace size calculation we also need to use `max(m,n)`

Reply:
as discussed offline, this is good solution but somehow for column major inputs there is perhaps fp arithmetic accuracy loss due to which I've kept `max(m,n)` only for row major inputs until we resolve it.